src/i2c.rs

```diff
@@ -0,0 +1,217 @@
+use std::marker::PhantomData;
+
+use crate::bindings::{
+    CgosI2CCount, CgosI2CGetFrequency, CgosI2CGetMaxFrequency, CgosI2CIsAvailable, CgosI2CRead,
+    CgosI2CReadRegister, CgosI2CSetFrequency, CgosI2CType, CgosI2CWrite, CgosI2CWriteReadCombined,
+    CgosI2CWriteRegister, CGOS_I2C_TYPE_DDC, CGOS_I2C_TYPE_PRIMARY, CGOS_I2C_TYPE_SMB,
+    CGOS_I2C_TYPE_UNKNOWN,
+};
+
+/// Error type for I2c operations
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum I2cErr {
+    /// User-supplied index is out of range
+    IdxOutOfRange,
+
+    /// I2c bus transaction failed
+    BusErr,
+}
+
+pub type I2cResult<T> = Result<T, I2cErr>;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum I2cType {
+    Unknown,
+    Primary,
+    Smb,
+    Ddc,
+    CongatecInternalUse(u32),
+}
+
+impl Into<u32> for I2cType {
+    fn into(self) -> u32 {
+        match self {
+            Self::Unknown => CGOS_I2C_TYPE_UNKNOWN,
+            Self::Primary => CGOS_I2C_TYPE_PRIMARY,
+            Self::Smb => CGOS_I2C_TYPE_SMB,
+            Self::Ddc => CGOS_I2C_TYPE_DDC,
+            Self::CongatecInternalUse(x) => x,
+        }
+    }
+}
+
+impl From<u32> for I2cType {
+    //note: On my devboard libcgos does return undeclared values for some busses, hence the need to return an error instead of panic
+    fn from(value: u32) -> I2cType {
+        match value {
+            CGOS_I2C_TYPE_UNKNOWN => Self::Unknown,
+            CGOS_I2C_TYPE_PRIMARY => Self::Primary,
+            CGOS_I2C_TYPE_SMB => Self::Smb,
+            CGOS_I2C_TYPE_DDC => Self::Ddc,
+            _ => Self::CongatecInternalUse(value),
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct I2c<'library> {
+    handle: u32,
+    index: u32,
+    _library_lifetime: PhantomData<&'library ()>,
+}
+impl<'library> I2c<'library> {
```

Comment:
This lifetime does not need to be named explicitly and can be elided.

```suggestion
impl I2c<'_> {
```